### ISSUE TITLE
"ImportError: No module named conda"

### ISSUE THREAD

**Issue by pitrou** [conda/conda#9507]
I recently upgrade conda from 4.7.12 to 4.8.0 and now, every time I open a shell, I get this error:
```
Traceback (most recent call last):
  File "<stdin>", line 3, in <module>
ImportError: No module named conda
```

Weirdly, conda still works:
```
(base) antoine@orphee:~$ conda info

     active environment : base
    active env location : /home/antoine/miniconda3
            shell level : 1
       user config file : /home/antoine/.condarc
 populated config files : 
          conda version : 4.8.0
    conda-build version : not installed
         python version : 3.7.5.final.0
       virtual packages : __glibc=2.29
       base environment : /home/antoine/miniconda3  (writable)
           channel URLs : https://repo.anaconda.com/pkgs/main/linux-64
                          https://repo.anaconda.com/pkgs/main/noarch
                          https://repo.anaconda.com/pkgs/r/linux-64
                          https://repo.anaconda.com/pkgs/r/noarch
          package cache : /home/antoine/miniconda3/pkgs
                          /home/antoine/.conda/pkgs
       envs directories : /home/antoine/miniconda3/envs
                          /home/antoine/.conda/envs
               platform : linux-64
             user-agent : conda/4.8.0 requests/2.22.0 CPython/3.7.5 Linux/5.0.0-37-generic ubuntu/19.04 glibc/2.29
                UID:GID : 1000:1000
             netrc file : None
           offline mode : False

```

If I remove the snippet added by `conda init` at the end of my `.bashrc`, the error disappears (but of course conda is not in the PATH anymore). If I run `conda init` again, the error reappears.

**Comment by pitrou** [comment]
This is the snippet originally added by `conda init` to `.bashrc`:
```bash
# >>> conda initialize >>>
# !! Contents within this block are managed by 'conda init' !!
__conda_setup="$('/home/antoine/miniconda3/bin/conda' 'shell.bash' 'hook' 2> /dev/null)"
if [ $? -eq 0 ]; then
    eval "$__conda_setup"
else
    if [ -f "/home/antoine/miniconda3/etc/profile.d/conda.sh" ]; then
        . "/home/antoine/miniconda3/etc/profile.d/conda.sh"
    else
        export PATH="/home/antoine/miniconda3/bin:$PATH"
    fi
fi
unset __conda_setup
# <<< conda initialize <<<
```

If I comment out some of the lines thusly:
```bash
# >>> conda initialize >>>
# !! Contents within this block are managed by 'conda init' !!
#__conda_setup="$('/home/antoine/miniconda3/bin/conda' 'shell.bash' 'hook' 2> /dev/null)"
#if [ $? -eq 0 ]; then
#    eval "$__conda_setup"
#else
    if [ -f "/home/antoine/miniconda3/etc/profile.d/conda.sh" ]; then
        . "/home/antoine/miniconda3/etc/profile.d/conda.sh"
    else
        export PATH="/home/antoine/miniconda3/bin:$PATH"
    fi
#fi
#unset __conda_setup
# <<< conda initialize <<<
```
then the error disappears and the conda install still works.

**Comment by QuentinSoubeyran** [comment]
Hi there, I got the same error and investigated a bit. Short version of my findings at the end.

- This is indeed caused by upgrading conda from  `4.7.12` to `4.8`, tested after completely removing conda (using `anaconda-clean`, `rm -rf ~/anaconda3` and removing the conda init lines from `~/.bashrc`), re-installing conda with the installer and upgrading conda.
    + I first tried `conda install pylint`, which only upgrades `conda`. The error occures afterwards
    + I then tried `conda upgrade conda`. Two other packages are upgraded with conda, but the error still appears

- The problems comes from the `eval "$__conda_setup"` line of `.bashrc`. Using `echo`, I saw that the `__conda_setup` returned by the shell.bash hook differ vastly from conda `4.7.12` and conda `4.8`. Conda `4.8` sources `<conda_install>/etc/bash_completion.d/conda` which does not exist under conda `4.7.12`. In this file, the problematic lines seems to be:

```bash
# line 26
function __comp_conda_ensure_root() {
   # [...]
        CONDA_SOURCE=$(
            cat - <<'            EOF' | sed 's/^ *: //g' | python -
            : from __future__ import print_function
            : import os
            : import conda # <--  THERE
            : print(os.path.dirname(conda.__file__))
            EOF
        )
    # [...]
}
```

I added  a `raise` statement (print doesn't work) as follows:

```bash
# line 26
function __comp_conda_ensure_root() {
   # [...]
        CONDA_SOURCE=$(
            cat - <<'            EOF' | sed 's/^ *: //g' | python -
            : from __future__ import print_function
            : import os
           # ADDED LINES
            : import sys
            : raise ValueError("python: " + sys.version)
           # END ADDED LINES
            : import conda # <--  THERE
            : print(os.path.dirname(conda.__file__))
            EOF
        )
    # [...]
}
```

While this obviously halts the execution, it showed that the python version was 2.7 - which I'm pretty sure should not be the case, this is a system python version, not the anaconda one.

When looking at the `PATH` just before the function is called, here in the same file:

```bash
# Lline 100
# cache conda subcommand help lookups for the duration of the shell
unset __comp_conda_cache
declare -A __comp_conda_cache
echo "$PATH" # ADDED LINE
__comp_conda_ensure_root
```

and comparing with the `PATH` after the shell has started up, during that execution the`PATH` is *missing* `~/anaconda3/bin`, which is present after the shell has started.

**TL;DR**: The problem seems to be that the new file `~/anaconda3/etc/bash_completion.d/conda` added by `conda 3.8.0` is sourced before `~/anaconda3/bin` is prefixed to the `PATH`, so the python used is the system python which naturally lacks the conda module.

Hope this helps resolve the issue !

**Comment by jeanmonet** [comment]
Same here. The bug does not seem to affect any `conda` commands - all seems to work even though I get this error each time I log into shell (Ubuntu 18.04 LTS):

```
Traceback (most recent call last):
  File "<stdin>", line 3, in <module>
ImportError: No module named conda
Traceback (most recent call last):
  File "<stdin>", line 3, in <module>
ImportError: No module named conda
Traceback (most recent call last):
  File "<stdin>", line 3, in <module>
ImportError: No module named conda
```

**Comment by piglesias1** [comment]
I had the same issue after the conda update to 4.8.0. I was checking the init script and could reach the same conclusion as @QuentinSoubeyran. As the python version is from the system and not from the conda package, due to the path was not set before, I did a simple workaround:

change the line 36 of `<conda_install>/etc/bash_completion.d/conda` from:
```sh
        CONDA_SOURCE=$(
            cat - <<'            EOF' | sed 's/^ *: //g' | python -  # <<< Original
            : from __future__ import print_function
            : import os
            : import conda
            : print(os.path.dirname(conda.__file__))
            EOF
        )
```
to: 
```sh
        CONDA_SOURCE=$(
            cat - <<'            EOF' | sed 's/^ *: //g' | ${CONDA_PYTHON_EXE} -  # <<< Changed
            : from __future__ import print_function
            : import os
            : import conda
            : print(os.path.dirname(conda.__file__))
            EOF
        )
```
Now seems to be working normally and I don't have the error message when starting the shell.

**Comment by MMStojiljkovic** [comment]
I can report the same problem on Ubuntu 18.04. After downgrading back to conda 4.7.12, it works fine.

**Comment by forrestwaters** [comment]
Looks like this is fixed in master by #9421 but was not included in the 4.8 release

**Comment by Brinkhuis** [comment]
Issue seems solved by todayś update `conda update --all`

**Comment by obarak** [comment]
> Issue seems solved by todayś update `conda update --all`

`conda update conda` in your `base` environment is enough, don't have to update all other packages :-)
BTW, the new package is still tagged 4.8.0 only with build py37_1 instead of py37_0

**Comment by hmaarrfk** [comment]
That fix only works for people using conda-forge. but I guess using conda-forge is your choice.

**Comment by jjhelmus** [comment]
`defaults` also has a patched build.

**Comment by msarahan** [comment]
fixed in 4.8.1, which is on conda-canary.  Closing.

**Comment by Vivek1717** [comment]
I am still getting the error:
C:\ProgramData\Anaconda3\condabin>conda update conda
Traceback (most recent call last):
  File "C:\ProgramData\Anaconda3\Scripts\conda-script.py", line 11, in <module>
    from conda.cli import main
ModuleNotFoundError: No module named 'conda'

**Comment by QuentinSoubeyran** [comment]
Hi @Vivek1717 ,
From the looks of it, you are getting a different error. The original error was a curious print when opening a console, with conda still working properly.

You have the `conda update` conda command (and possibly others) crashing on execution.

I would suggest looking for your specific error online ([this](https://stackoverflow.com/questions/19825250/after-anaconda-installation-conda-command-fails-with-importerror-no-module-na) might help) and creating a dedicated issue if you still experience the problem.
